docs/understanding-airbyte/airbyte-protocol.md

```diff
@@ -0,0 +1,735 @@
+
+## Schema Mismatches
+
+Over time, it is possible for the catalog to become out of sync with the underlying data store it represents. The Protocol is design to be resilient to this. In should never fail due to a mismatch.
+
+|Scenario|Outcome|
+|---|---|
+|Stream exists in catalog but not in data store|When the source runs, it will not find the data for that stream. All other streams sync as usual.|
+|Stream exists in data store but not in catalog|When the source runs, it never looks for the data in the store related to that stream and thus does not emit it.
+|Field exists in catalog but not in data store|If the column for a table is remove in the underlying data store the Source will not find it and will not replicate it. It should not cause a failure. The data simply will not be there.|
+|Field exists in data store but not in catalog|When the source runs, it never looks for the field in the store. It should not emit that field. If it does, it should be ignored downstream. The existence of an unknown field should not cause a failure.|
+
+In short, if the catalog is ever out of sync with the schema of the underlying data store, it should never block replication for data that is present.
+
+# State & Checkpointing
+Sources are able to emit state in order to allow checkpointing data replication. The goal is that given wherever a sync stops (whether this is due to all data available at the time being replicated or due to a failure), the next time the Source attempts to extract data it can pick up where it left off and not have to go back to the beginning.
+
+This concept enables incremental syncs--syncs that only replicate data that is new since the previous sync.
+
+State also enables Partial Success. In the case where during a sync there is a failure before all data has been extracted and committed, if all records up to a certain state are committed, then the next time the sync happens, it can start from that state as opposed to going back to the beginning. Partial Success is powerful, because especially in the case of high data volumes and long syncs, being able to pick up from wherever the failure occurred can costly re-syncing of data that has already been replicated.
+
+## State & Source
+This section will step through how state is used to allow a Source to pick up where it left off. A Source takes state as an input. A Source should be able to take that input and use it to determine where it left off the last time. The contents of the Source is a black box to the Protocol. The Protocol provides an envelope for the Source to put its state in and then passes the state back in that envelope. The Protocol never needs to know anything about the contents of the state. Thus, the Source can track state however makes most sense to it.
+
+Here is an example of the lifecycle of state in reference to the Source.
+![](../.gitbook/assets/source-state-example.png)
```

Comment:
I like the images, specifically docs/.gitbook/assets/source-state-example.png!  Could you add a note on the arrow pointing to the next sync that shows that time has passed, and the drew row in now there, e.g.:  "pass last state to next sync - new rows since previous sync".  This is clear in the text content, but if this image is shared without the doc, it could be more clear.

Reply:
great! added.